quality-assurance/core.md

```diff
@@ -0,0 +1,192 @@
+
+**Login &amp; Auth**
+* Users can only access pages they are permitted access to.
+* Authentication when they are logged in
+* Login, registration (if applicable) and logout works as expected.
+* Reset password works as expected
+
+**Registration**
+* Email addresses are verified &amp; unique (if applicable)
+* Passwords meet minimum standards (min 8 chars)
+* Passwords should not have maximum length restrictions (allow at least 255 characters).
+
+**Emails**
+* Emails look good on small form factor devices
+* Ensure correct subject &amp; sender name
+* Correct email is sent, information is conveyed appropriately
+* Emails are received in a timely manner
+* Emails come from the appropriate email (often the clients)
+* Links point to the correct URIs
+
+**Push Notifications**
+* Push notifications are received in a timely manner
+* They show correct &amp; relevant information
+
+**Permissions Roles**
+* Ensure permissions &amp; roles match up to possible actions (only admins can create users etc)
+
+### Whitebox
+**General**
+* Filenames must follow standard coding conventions (UserController.php)
+* Classes must follow standard coding conventions (UserController)
+* Variables should follow camelCase &amp; be descriptive (myName)
+* All long running operations (at your discretion) are queued
+* APIs follow REST architecture as closely as possible and implement our [API Response](https://github.com/langleyfoxall/helpers-laravel#apiresponse)
+* Auth middleware is setup &amp; working correctly
+* Event system is used for method side effects (Audit trail etc)
+* Laravel Notifications are used to notify the user of system events (subscription expired, new friend request etc)
+* Sensitive values are either hashed or encrypted (when required by law or common sense)
+* Third party API integrations that can &amp; will be re-used across projects should be implemented as external packages, 
+assuming no existing open-source integration exists, to allow for independent testing and code reuse if appropriate.
+* Ensure third party API keys (e.g. SES keys) are not present within .env.example and other relevant files outside of the 
+production / development environments.
+* If Travis CI is being used, ensure third party API keys are input directly into Travis CI's website relevant to the 
+project and make sure "Display value in build log" is not checked.
+* Log viewers are hidden behind a protected route
+
+**Controllers**
+* Be on the lookout for duplicated code - recommend refactoring into helper classes or models if so
+* Ensure all requests are validated either through a validator or custom request
+* Errors are handled correctly in try catch blocks
+* Responses are returned in a consistent &amp; logical format
+* Only appropriate &amp; relevant information is returned
+* When returning JSON ensure the returned JSON conforms to the [JSON spec](http://json-schema.org/specification.html)
+* Appropriate status codes are set
```

Comment:
```suggestion
* Appropriate status codes are set
  * `201` - For creation API responses
  * `202` - For sending requests that are successful to a third party API
  * `401` - For unauthenticated users trying to access protected areas
  * `403` - For normal users trying to access admin areas
  * `422` - For missing required data in the request body
  * Continue reading [here](https://httpstatuses.com/)
```

Might it be worth giving examples or linking off to a resource with descriptions on HTTP status codes? I personally like [httpstatuses.com](https://httpstatuses.com/).

Discussing what status codes to be used where/when is probably something that can defined in more detail in another document.

Reply:
Interesting that you've missed `204` and `200` off there as common ones.

Unsure whether this is the right document for them - thumbs 👍 👎

Reply:
I think having examples of the right status codes is the right way, enforcing them I think puts a blanket requirement on something that isn't always black and white.

What is we're talking to a third party and then creating a resource, would we go with `201`, `202` or just plain old `200`?

I think it needs it's own discussion about the correct status codes to use in different circumstances.

Reply:
> Unsure whether this is the right document for them - thumbs 👍 👎

> I think it needs it's own discussion about the correct status codes to use in different circumstances.

As mentioned, I think that status codes could be it's own discussion and document but we should at least point towards a good resource for status codes, as I'm sure some people wouldn't bother googling and just put `200`s for everything.

Reply:
I think example should be pulled into their own markdown file, linked to from here.

Reply:
Executive decision to leave this for another section of the tech repo. This doesn't quite relate to Quality Assurance enough for me.